src/Microsoft.Framework.OptionsModel/ConfigurationBinder.cs

```diff
@@ -0,0 +1,284 @@
+            var valueType = iDictionaryTypeInfo.GenericTypeArguments[1];
+
+            if (keyType != typeof(string))
+            {
+                // We only support string keys
+                return;
+            }
+
+            var addMethod = iDictionaryTypeInfo.GetDeclaredMethod("Add");
+            var subkeys = _currentConfigurationContext.GetSubKeys().ToList();
+
+            foreach (var keyProperty in subkeys)
+            {
+                EnterContext(keyProperty.Value);
+
+                try
+                {
+                    var item = BindType(valueType, null);
+                    if (item != null)
+                    {
+                        addMethod.Invoke(dictionary, new[] { keyProperty.Key, item });
+                    }
+                }
+                catch
+                {
+                }
+                finally
+                {
+                    ExitContext();
+                }
+            }
+        }
+
+        private void BindCollection(object collection, Type iCollectionType)
+        {
+            var iCollectionTypeInfo = iCollectionType.GetTypeInfo();
+
+            // It is guaranteed to have a one and only one parameter
+            // because this is an ICollection<T>
+            var itemType = iCollectionTypeInfo.GenericTypeArguments[0];
+
+            var addMethod = iCollectionTypeInfo.GetDeclaredMethod("Add");
+            var subkeys = _currentConfigurationContext.GetSubKeys().ToList();
+
+            foreach (var keyProperty in subkeys)
+            {
+                EnterContext(keyProperty.Value);
+
+                try
+                {
+                    var item = BindType(itemType, null);
+                    if (item != null)
+                    {
+                        addMethod.Invoke(collection, new[] { item });
+                    }
+                }
+                catch
+                {
+                }
+                finally
+                {
+                    ExitContext();
+                }
+            }
+        }
+
+        private object CreateValueFromConfiguration(Type type, string value)
+        {
+            var typeInfo = type.GetTypeInfo();
+
+            try
+            {
+                if (typeInfo.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>))
+                {
+                    return CreateValueFromConfiguration(Nullable.GetUnderlyingType(type), value);
+                }
+                else if (typeInfo.IsEnum)
+                {
+                    return Enum.Parse(type, _currentConfigurationContext.Get(null));
+                }
+                else
+                {
+                    return Convert.ChangeType(_currentConfigurationContext.Get(null), type);
+                }
+            }
+            catch
+            {
+                return null;
+            }
+        }
+
+        private void EnterContext(IConfiguration context)
```

Comment:
So looking at this code, do you really need a stack of IConfiguration?  Can all of the helpers be made static with the _currentConfigurationContext passed via a parameter instead?  I think it reads a bit easier overall without the Stack of IConfigurations unless they are actually needed for a specific reason...